test/lint_cases.yaml

```diff
@@ -219,6 +219,17 @@ tests:
         sha256: 123
       - url: https://elsewhere  # [osx]
         sha256: 123
+- name: selector_in_source_with_compiler
+  remove:
+    - source
+    - build/noarch
+  add:
+    source:
+      - url: https://somewhere  # [linux]
+        sha256: 123
+      - url: https://elsewhere  # [osx]
+        sha256: 123
+    requirements: { build: ['{{compiler("c")}}'] }
```

Comment:
If we have a recipe that doesn't use `{{ compiler(...) }}` but has selectors in `source`, `should_be_noarch_generic` is still triggered. We likely wouldn't want such recipe to be `noarch`, but I'm not too concerned about that since that would be rather uncommon => I'd say let's just fix/handle that if we encounter it :man_shrugging:.

Reply:
Shouldn't the `sha`s have selectors too? Otherwise you have duplicate keys after rendering.

Reply:
Yep, I just mindlessly copy-pasted that part.

Reply:
-> gh-627